csrc/kernels.cu

```diff
@@ -219,7 +220,7 @@ __device__ half dhDequantizeNF4(unsigned char val)
 
 }
 
-__device__ float dDequantizeNF4(unsigned char val)
+__device__ __forceinline__ float dDequantizeNF4(unsigned char val)
```

Comment:
I suppose this `__forceinline__` has been measured to be Always A Good Idea?

Reply:
It's not always a good idea, but in this case I saw a small throughput improvement. The same improvement may exist for FP4 as well, but I did not test.

Reply:
That's something I was thinking about when I first started looking at these too. The fast GEMV kernel `kgemm_4bit_inference_naive` actually does it that way - the codebook for the datatype is loaded from global memory to shared. While the `kDequantizeBlockwise` kernel also has an argument for a code table, it's not used, and instead this hard-coded dequantization function is used instead.

I haven't spent much time profiling this but definitely share that curiosity. With that said, the compiler might be smart enough here to avoid divergence with predication or some other tricks.